File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.java
import java.util.*;
import java.lang.*;
import java.io.*;

public class Main {
	static BufferedReader br;
	static BufferedWriter bw;
	static StringTokenizer st;

    @SuppressWarnings("unchecked")
	public static void main(String[] args) throws Exception {

		br = new BufferedReader(new InputStreamReader(System.in));
		bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        List<City>[] adj = new List[n+1];
        int[] dist = new int[n+1];
        for (int i = 0; i <= n; i++) {
            adj[i] = new ArrayList<>();
            dist[i] = Integer.MAX_VALUE;
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());

            adj[start].add(new City(end, cost));
        }

        String[] input = br.readLine().split(" ");
        int start = Integer.parseInt(input[0]);
        int dest = Integer.parseInt(input[1]);

        PriorityQueue<City> pq = new PriorityQueue<>();
        pq.add(new City(start, 0));
        dist[start] = 0;
        int answer = 0;

        while (!pq.isEmpty()) {

            City curr = pq.poll();
            if (curr.num == dest) {
                answer = curr.dist;
                break;
            }

            for (City city : adj[curr.num]) {
                if (dist[city.num] > curr.dist + city.dist) {
                    dist[city.num] = curr.dist + city.dist;
                    pq.add(new City(city.num, dist[city.num]));
                }
            }
        }

        bw.write(String.valueOf(answer));
		
		bw.flush();
		bw.close();
		br.close();
	}
}

class City implements Comparable<City> {
    int num;
    int dist;

    City () {}
    City (int num, int dist) {
        this.num = num;
        this.dist = dist;
    }

    @Override
    public int compareTo(City o) {
        return Integer.compare(this.dist, o.dist);
    }
}

File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <sstream>
#include <climits>
#include <queue>
using namespace std;

class City {
    public:
        int num, dist;

        City () : num(-1), dist(INT_MAX) {}
        City (int a, int b) : num(a), dist(b) {}
};

struct compareDist {
    bool operator() (const City& c1, const City& c2) {
        return c1.dist > c2.dist;
    }
};

int main() {

    int n, m;
    cin >> n >> m;
    cin.ignore();

    vector<vector<City> > adj(n+1);
    
    for (int i = 0; i < m; i++) {
        string input;
        getline(cin, input);
        stringstream ss(input);
        int start, end, dist;
        ss >> start >> end >> dist;
        adj[start].push_back(City(end, dist));
    }

    int start, end;
    cin >> start >> end;

    vector<int> dp(n+1);
    fill(dp.begin(), dp.end(), INT_MAX);
    priority_queue<City, vector<City>, compareDist> pq;
    pq.push(City(start, 0));
    dp[start] = 0;

    while(!pq.empty()) {

        City curr = pq.top();
        pq.pop();

        if (curr.num == end) break;

        for (City city : adj[curr.num]) {
            if (dp[city.num] > curr.dist + city.dist) {
                dp[city.num] = curr.dist + city.dist;
                pq.push(City(city.num, dp[city.num]));
            }
        }
    }

    cout << dp[end];

    return 0;
}